### PR TITLE
Add fbx autodetection

### PR DESCRIPTION
**Pull request by foreign-sub**
Detect freebox host, port and api_version
cleanup, format, refactor
Support any number of freeboxes
Open by uid
breaking change : token file has been removed, each freebox now has it's own token file indexed by uid
api modules instances are now created only when they're called.
settings can be stored anywhere using the data_dir option.

### CODE REVIEW

File: aiofreepybox/aiofreepybox.py
Comment (SNoof85) [inline aiofreepybox/aiofreepybox.py]: Not sure we want to do this if values are hardcoded (not set to auto)

Comment (foreign-sub) [reply aiofreepybox/aiofreepybox.py]: You're right i also figured it was not ideal, it should be fine now.

File: aiofreepybox/aiofreepybox.py
Comment (stilllman) [inline aiofreepybox/aiofreepybox.py]: Maybe split on `.` and get the first token rather than expecting the version to fit on a single digit (I assume version 10.0 will happen sooner or later :) )

Comment (foreign-sub) [reply aiofreepybox/aiofreepybox.py]: It could happen around 2023, i'll split it.

**Comment by stilllman** [comment]
@SNoof85 Any opinion on this change? If I remember correctly, the API version is fixed in Home Assistant to v4, so the change should not require any changes. It could probably be used to simplify the hass configuration.

**Comment by SNoof85** [comment]
For sure in home assistant the api version is fixed in the Freebox integration : https://github.com/home-assistant/home-assistant/blob/71f73af535f68f8df8b4f9e229942942c67a46d3/homeassistant/components/freebox/__init__.py#L63

So in my opinion it won't break anything and i like this feature. Problem is still lack of documentation and i need to test with latest API what we have already in home assistant before using this. (if i see any use case...).

Looks good to me for merging ! 🚀

**Comment by foreign-sub** [comment]
Since the latest release (4.1) the API version is now 7.

**Comment by SNoof85** [comment]
This release has not hit my Freebox révolution right now I can't do any testing. We must keep an option to make the calls with a specific API version. Autodetect can be good to have but it should be possible to override.

**Comment by stilllman** [comment]
Hey @foreign-sub! Thanks a lot for all the work you put in this library, really, but honestly I don't know how I am supposed to review your PRs: every time I come back on a PR the entire code has changed since the last time I reviewed, so I need to review everything from scratch. I'm not that attached to this library, so I'm not sure I want to spend dozens of evening reviewing code for it (I have a family, and I already do code reviews all day long at work 😄). Moreover, you seem to have more information than me about the Freebox API, the only thing I have is the doc of API v4 [here](https://dev.freebox.fr/sdk/os/), and I only have a Freebox mini 4K. What would you think of continuing development on your own fork? I would transfer ownership of the pypi package to you (and @SNoof85 if he wants to?) so that you can publish new packages. What do you say?

**Comment by foreign-sub** [comment]
Hi, @stilllman, sorry about that, it was totally unforeseen. If you had asked me about two months ago, my only plan was to add support for the home API, then things kind of came up one after another. I don't have access to anymore doc than you do, i just scrap whatever data i can find around the web, but it involves mostly browsing and reviewing source code whatever the language may be.
I don't really know if i'm ready to continue on my own, i guess that if you're asking you kind of feel i am (and also want to get rid of my daily push :)), but i want you to know that your contribution has been decisive into making the API what it is now, and that i'd be sad to lose the technical expertise and patience you provided when reviewing my code. Now i totally understand your motivation, would it be acceptable to commit the remainings pr's, publish one last release, then transfer the ownership afterwards to me and @SNoof85 of course if he want's to, i think i got one last small commit on this branch then i'm done, unless some bugs needs to be fixed of course.

**Comment by SNoof85** [comment]
I think we can let some time to @foreign-sub to polish and then make a final review, publish a package.
Then @stilllman if you want to transfer the ownership it's not an issue for me.

**Comment by stilllman** [comment]
Sorry for the long silence, I'll try to finalize the reviews during the christmas holidays

**Comment by stilllman** [comment]
Oh! @SNoof85 @foreign-sub I just found by accident that the up-to-date API documentation is available in the Freebox help menu, on http://mafreebox.freebox.fr !

**Comment by SNoof85** [comment]
No way ! How could we missed that for ages !
That is a real good finding !

**Comment by foreign-sub** [comment]
Hey @stilllman , @SNoof85 ,

Sorry it took so long, typing was ..., but it's done.

I've merged everything in the following branch if you wanna give it a try:
https://github.com/foreign-sub/aiofreepybox/tree/afpbx-next
I've left out #50 for now, I still need to review it, but it should not be an issue.

File: aiofreepybox/aiofreepybox.py
Comment (stilllman) [inline aiofreepybox/aiofreepybox.py]: Wow, we went from a 50-lines auto-discovery to something that is around 400/500 lines, what happened?! The first version seemed to work, was it incorrect? I tried understanding the new stuff you did with the various disc_* functions, the db, etc., but it's too complex for me, I can't get my head around it... I'll come back to it with a fresher eye, but in the meantime I reiterate my proposal to simply transfer the project ownership to you if you want to move forward without waiting for my approval.

File: aiofreepybox/aiofreepybox.py
Comment (foreign-sub) [reply aiofreepybox/aiofreepybox.py]: Fixed.

File: aiofreepybox/aiofreepybox.py
Comment (foreign-sub) [reply aiofreepybox/aiofreepybox.py]: Yes, it has some draw backs (ie autocompletion issues mostly), but it also reduces the startup time and the memory footprint.

File: aiofreepybox/aiofreepybox.py
Comment (foreign-sub) [reply aiofreepybox/aiofreepybox.py]: I had some weird issues sometimes when doing tests, after searching around i found out that in aiohttp graceful shutdown doc, they recommend allowing a small delay when closing an ssl connection.

> ... through a direct ClientSession.close() call), the underlying connection remains open due to asyncio internal details.  ... To avoid this situation, a small delay must be added before closing the event loop to allow any open underlying connections to close.

https://docs.aiohttp.org/en/stable/client_advanced.html

File: aiofreepybox/aiofreepybox.py
Comment (foreign-sub) [reply aiofreepybox/aiofreepybox.py]: Exactly, whatever the state (like losing the link before the end) it will always close without complaining, that's why i reduced the scope as much as i could (to AttributeError ) and added one try block per request.

File: aiofreepybox/aiofreepybox.py
Comment (foreign-sub) [reply aiofreepybox/aiofreepybox.py]: No worries, I'll explain in a few words (or maybe a bit more), it all started when @SNoof85 opened issue #52.
The first version was ok, but it had it's own limits, like the one @SNoof85 raised. Technically this issue (#52) had been here long before, but it was up to the user to figure out a way to workaround it (which was both kind of painful and complicated). This is the result, and it came with lots of perks, first ones being, the uid and the db, now you can manage a whole fleet of freeboxes easily, no more switching token files and making sure you've set the correct ip adress, now once setup is done you're good to go.
As i was writing it, it turned out to be complex and unreadable.
Here is a quick description of what it does so you can get around it quickly:

discover:
It still tries to find a freebox on the lan after checking it has not opened a session already, then updates the db if it found something.

open:
Now, first it will try to find the uid of the freebox you're requesting if it was not provided (uid is now the preferred way to adress any freebox) but it still supports the "host" "port" format,  otherwise it will load the db for the specified uid.
Then as before it will try to open a session, but unlike before, it will try every way it knows to contact the freebox until it fails if the freebox can't be reached.

Each freebox creates two files, the app_auth (now named ".fbx_app_auth_{uid_base64}" ) like before and the db (named " .fbx_db_{uid_base64}") those can now be stored wherever you want using the "data_dir" option, they can be read with bzcat.
The db store known freebox adresses, linked with the fbx_desc. It can be managed with get_db and clean_db.

The rest is made up of support functions for discover and open as they can combine in some  ways and db management.

**Comment by SNoof85** [comment]
@stilllman for me it's OK to get the ownership 🎉

File: aiofreepybox/aiofreepybox.py
Comment (stilllman) [reply aiofreepybox/aiofreepybox.py]: Thanks for the thorough explanation!

File: aiofreepybox/aiofreepybox.py
Comment (stilllman) [reply aiofreepybox/aiofreepybox.py]: Very interesting, thanks! I keep learning new things with your PRs 😄 !

File: .gitignore
Comment (Quentame) [inline .gitignore:108]: Should have a EOF

File: aiofreepybox/aiofreepybox.py
Comment (Quentame) [inline aiofreepybox/aiofreepybox.py:884]:
```suggestion
    def fbx_api_modules(self) -> Optional[list]:
```

**Review by Quentame — CHANGES_REQUESTED** [review]
Not finished review

**Pull request closed without merging**
